[rpc] merge and separate out RRef python definition

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42944 [WIP][c10d][jit] torchbind ProcessGroup
* #44046 [c10d] switch ProcessGroup::Work to be managed by intrusive_ptr
* #44492 [c10] make intrusive_ptr available as a pybind holder type
* #44045 [c10] add operator= of intrusive_ptr to weak_intrusive_ptr
* #44044 [torchbind] add static define for C++ method
* **#42751 [rpc] merge and separate out RRef python definition**

This PR separate out RRef python definition by merging the Jit type
detection and RRef/Typing, and separate it out to a separate file to
make it clear.

Also do conditional imports to resolve any circular import we might
introduce in the future

Differential Revision: [D23228683](https://our.internmc.facebook.com/intern/diff/D23228683)

wanchaol committed:
[rpc] merge and separate out RRef python definition

This PR separate out RRef python definition by merging the Jit type
detection and RRef/Typing, and separate it out to a separate file to
make it clear.

Also do conditional imports to resolve any circular import we might
introduce in the future

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] merge and separate out RRef python definition"


This PR separate out RRef python definition by merging the Jit type
detection and RRef/Typing, and separate it out to a separate file to
make it clear.

Also do conditional imports to resolve any circular import we might
introduce in the future

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] merge and separate out RRef python definition"


This PR separate out RRef python definition by merging the Jit type
detection and RRef/Typing, and separate it out to a separate file to
make it clear.

Also do conditional imports to resolve any circular import we might
introduce in the future

[ghstack-poisoned]

mrshenli commented on torch/_rref.py:
```diff
@@ -0,0 +1,72 @@
+# This is for the fact that pybind11 generates the parameter
+# `self` as type `rpc.PyRRef`, so a `:inherited-members:`
+# under `.. autoclass:: RRef` does not work.
+# we have to do the following process to replacee `rpc.PyRRef` with `rpc.RRef`.
```
replacee -> replace

mrshenli commented on torch/jit/annotations.py:
```diff
@@ -49,7 +44,8 @@ class EvalEnv(object):
     def __init__(self, rcb):
         self.rcb = rcb
-        if torch.distributed.rpc.is_available():
+        if torch.distributed.is_available():
```
any reason for just checking distributed instead of RPC? They should be the same for now, but we might change build script to have dedicated flag for RPC later?

mrshenli reviewed (commented):
Changes LGTM. Test failures are real. BTW, could you please build the doc to verify that RRef docs are as expected?

```
Aug 18 22:12:05 ======================================================================
Aug 18 22:12:05 FAIL [54.072s]: test_run_mypy (__main__.TestTypeHints)
Aug 18 22:12:05 Runs mypy over all files specified in mypy.ini
Aug 18 22:12:05 ----------------------------------------------------------------------
Aug 18 22:12:05 Traceback (most recent call last):
Aug 18 22:12:05   File "test_type_hints.py", line 218, in test_run_mypy
Aug 18 22:12:05     self.fail("mypy failed: {}".format(stdout))
Aug 18 22:12:05 AssertionError: mypy failed: torch/_rref.py:4: error: Module 'torch.distributed.rpc' has no attribute 'PyRRef'; maybe "RRef"?  [attr-defined]
Aug 18 22:12:05 torch/_rref.py:13: error: Variable "torch._rref.RRefTypeVar" is not valid as a type  [valid-type]
Aug 18 22:12:05 torch/_rref.py:13: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
Aug 18 22:12:05 torch/_rref.py:13: error: Invalid base class "RRefTypeVar"  [misc]
Aug 18 22:12:05 torch/_rref.py:18: error: Name 'RRefTypeVar.__class__' is not defined  [name-defined]
Aug 18 22:12:05 torch/_rref.py:22: error: Name 'RRef' already defined on line 13  [no-redef]
Aug 18 22:12:05 torch/_rref.py:22: error: Variable "torch._rref.RRefTypeVar" is not valid as a type  [valid-type]
Aug 18 22:12:05 torch/_rref.py:22: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
Aug 18 22:12:05 torch/_rref.py:22: error: Invalid base class "RRefTypeVar"  [misc]
Aug 18 22:12:05 Found 7 errors in 1 file (checked 1053 source files)
```

wanchaol committed:
Update on "[rpc] merge and separate out RRef python definition"


This PR separate out RRef python definition by merging the Jit type
detection and RRef/Typing, and separate it out to a separate file to
make it clear.

Also do conditional imports to resolve any circular import we might
introduce in the future

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] merge and separate out RRef python definition"


This PR separate out RRef python definition by merging the Jit type
detection and RRef/Typing, and separate it out to a separate file to
make it clear.

Also do conditional imports to resolve any circular import we might
introduce in the future

Differential Revision: [D23228683](https://our.internmc.facebook.com/intern/diff/D23228683)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] merge and separate out RRef python definition"


This PR separate out RRef python definition by merging the Jit type
detection and RRef/Typing, and separate it out to a separate file to
make it clear.

Also do conditional imports to resolve any circular import we might
introduce in the future

Differential Revision: [D23228683](https://our.internmc.facebook.com/intern/diff/D23228683)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] merge and separate out RRef python definition"


This PR separate out RRef python definition by merging the Jit type
detection and RRef/Typing, and separate it out to a separate file to
make it clear.

Also do conditional imports to resolve any circular import we might
introduce in the future

Differential Revision: [D23228683](https://our.internmc.facebook.com/intern/diff/D23228683)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] merge and separate out RRef python definition"


This PR separate out RRef python definition by merging the Jit type
detection and RRef/Typing, and separate it out to a separate file to
make it clear.

Also do conditional imports to resolve any circular import we might
introduce in the future

Differential Revision: [D23228683](https://our.internmc.facebook.com/intern/diff/D23228683)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] merge and separate out RRef python definition"


This PR separate out RRef python definition by merging the Jit type
detection and RRef/Typing, and separate it out to a separate file to
make it clear.

Also do conditional imports to resolve any circular import we might
introduce in the future

Differential Revision: [D23228683](https://our.internmc.facebook.com/intern/diff/D23228683)

[ghstack-poisoned]
